### PR TITLE
[api] separate grpcserver from server

### PR DESCRIPTION
**Pull request by Liuhaai**
Code reconstruction: separate grpcserver from `api.go` for better code structure

linked #2866

### CODE REVIEW

File: api/grpcserver.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+	log.L().Info("grpc server is listening.", zap.String("addr", lis.Addr().String()))
+	go func() {
+		if err := api.grpcServer.Serve(lis); err != nil {
+			log.L().Fatal("grpc failed to serve.", zap.Error(err))
+		}
+	}()
+	return nil
+}
+
+func (api *GrpcServer) Stop() {
```
Comment (CoderZhi) [inline api/grpcserver.go:75]: `func (api *GrpcServer) Stop() error {`

File: api/grpcserver.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+// GetAccount returns the metadata of an account
+func (svr *GrpcServer) GetAccount(ctx context.Context, in *iotexapi.GetAccountRequest) (*iotexapi.GetAccountResponse, error) {
+	accountMeta, blockIdentifier, err := svr.coreService.GetAccount(in.Address)
+	if err != nil {
+		return nil, err
+	}
+	return &iotexapi.GetAccountResponse{
+			AccountMeta:     accountMeta,
+			BlockIdentifier: blockIdentifier},
+		nil
```
Comment (CoderZhi) [inline api/grpcserver.go:97]: format

File: api/grpcserver.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	chainMeta, syncStatus, err := svr.coreService.GetChainMeta()
+	if err != nil {
+		return nil, err
+	}
+	return &iotexapi.GetChainMetaResponse{ChainMeta: chainMeta, SyncStage: syncStatus}, nil
+}
+
+// GetServerMeta gets the server metadata
+func (svr *GrpcServer) GetServerMeta(ctx context.Context, in *iotexapi.GetServerMetaRequest) (*iotexapi.GetServerMetaResponse, error) {
+	return svr.coreService.GetServerMeta()
```
Comment (CoderZhi) [inline api/grpcserver.go:173]: why coreService return iotexapi.*?

File: api/grpcserver.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+		ReceiptInfo: &iotexapi.ReceiptInfo{
+			Receipt: receipt.ConvertToReceiptPb(),
+			BlkHash: blkHash,
+		},
+	}, nil
+}
+
+// ReadContract reads the state in a contract address specified by the slot
+func (svr *GrpcServer) ReadContract(ctx context.Context, in *iotexapi.ReadContractRequest) (*iotexapi.ReadContractResponse, error) {
+	return svr.coreService.ReadContract(in.Execution, in.CallerAddress, in.GasLimit, ctx)
```
Comment (CoderZhi) [inline api/grpcserver.go:205]: same question here

File: api/grpcserver.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	ret, err := svr.coreService.EstimateActionGasConsumption(ctx, in)
+	if err != nil {
+		return nil, err
+	}
+	return &iotexapi.EstimateActionGasConsumptionResponse{Gas: ret}, nil
+}
+
+// GetEpochMeta gets epoch metadata
+func (svr *GrpcServer) GetEpochMeta(ctx context.Context, in *iotexapi.GetEpochMetaRequest) (*iotexapi.GetEpochMetaResponse, error) {
+	return svr.coreService.GetEpochMeta(in.EpochNumber)
```
Comment (CoderZhi) [inline api/grpcserver.go:240]: and here

File: api/serverV2.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/iotexproject/iotex-core/blockindex"
+	"github.com/iotexproject/iotex-core/blocksync"
+	"github.com/iotexproject/iotex-core/config"
+	"github.com/iotexproject/iotex-core/state/factory"
+)
+
+type ServerV2 struct {
+	core       *CoreService
+	grpcServer *GrpcServer
+	httpServer *HttpServer
```
Comment (CoderZhi) [inline api/serverV2.go:17]: web3Server

File: api/serverV2.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func (svr *ServerV2) Start() error {
+	if err := svr.core.Start(); err != nil {
+		return err
+	}
+	if err := svr.grpcServer.Start(); err != nil {
+		return err
+	}
+	svr.httpServer.Start()
```
Comment (CoderZhi) [inline api/serverV2.go:50]: delete for now

File: api/serverV2.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	return nil
+}
+
+func (svr *ServerV2) Stop() error {
+	if err := svr.httpServer.Stop(); err != nil {
```
Comment (CoderZhi) [inline api/serverV2.go:55]: delete for now

File: api/coreservice.go [hunk 9]
```diff
@@ -0,0 +1,1608 @@
...
+// // WithNativeElection is the option to return native election data through API.
+// func WithNativeElection(committee committee.Committee) Option {
+// 	return func(cfg *Config) error {
+// 		cfg.electionCommittee = committee
+// 		return nil
+// 	}
+// }
+
+// Server provides api for user to query blockchain data
+type CoreService struct {
```
Comment (CoderZhi) [inline api/coreservice.go:95]: `coreService`

File: api/coreservice.go [hunk 10]
```diff
@@ -0,0 +1,1608 @@
...
+	cfg               config.Config
+	registry          *protocol.Registry
+	chainListener     Listener
+	hasActionIndex    bool
+	electionCommittee committee.Committee
+	readCache         *ReadCache
+}
+
+// NewServer creates a new server
+func NewCoreService(
```
Comment (CoderZhi) [inline api/coreservice.go:114]: `newCoreService`

File: api/coreservice.go [hunk 11]
```diff
@@ -0,0 +1,1608 @@
...
+		readCache:         NewReadCache(),
+	}
+	if _, ok := cfg.Plugins[config.GatewayPlugin]; ok {
+		svr.hasActionIndex = true
+	}
+	return svr, nil
+}
+
+// GetAccount returns the metadata of an account
+func (api *CoreService) GetAccount(addrStr string) (*iotextypes.AccountMeta, *iotextypes.BlockIdentifier, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:164]: Delete "Get" and rename it to Account.

File: api/coreservice.go [hunk 12]
```diff
@@ -0,0 +1,1608 @@
...
+		BuildTime:       version.BuildTime,
+	}}, nil
+}
+
+// SendAction is the API to send an action to blockchain.
+func (api *CoreService) SendAction(in *iotextypes.Action, chainID uint32, ctx context.Context) (string, error) {
+	log.L().Debug("receive send action request")
+	var selp action.SealedEnvelope
+	var err error
+	if err = selp.LoadProto(in); err != nil {
```
Comment (CoderZhi) [inline api/coreservice.go:304]: `if err := selp.LoadProto(in); err != nil {`

File: api/coreservice.go (hunk 11 above)
Comment (Liuhaai) [reply api/coreservice.go:164]: incompatible with current proto file

File: api/serverV2.go (hunk 6 above)
Comment (dustinxie) [reply api/serverV2.go:17]: i think remove `httpServer` in this PR?
in next PR, add it and the babel implementation

File: api/serverV2.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/iotexproject/iotex-core/state/factory"
+)
+
+type ServerV2 struct {
+	core       *CoreService
+	grpcServer *GrpcServer
+	httpServer *HttpServer
+}
+
+func NewServerV2(
```
Comment (dustinxie) [inline api/serverV2.go:20]: need to use it in `chainservice.go`?

Comment (Liuhaai) [reply api/serverV2.go:20]: The new code will be integrated into `chainservice.go` in future PR

File: api/coreservice.go [hunk 14]
```diff
@@ -681,7 +674,7 @@ func (api *CoreService) GetLogs(in *iotexapi.GetLogsRequest) ([]*iotextypes.Log,
 }
 
 // StreamBlocks streams blocks
-func (api *CoreService) StreamBlocks(stream iotexapi.APIService_StreamBlocksServer) error {
+func (api *coreService) StreamBlocks(stream iotexapi.APIService_StreamBlocksServer) error {
```
Comment (CoderZhi) [inline api/coreservice.go:677]: we need to update the input time as well, but we can leave it to follow up PR

File: api/coreservice.go [hunk 15]
```diff
@@ -699,7 +692,7 @@ func (api *CoreService) StreamBlocks(stream iotexapi.APIService_StreamBlocksServ
 }
 
 // StreamLogs streams logs that match the filter condition
-func (api *CoreService) StreamLogs(in *iotexapi.LogsFilter, stream iotexapi.APIService_StreamLogsServer) error {
+func (api *coreService) StreamLogs(in *iotexapi.LogsFilter, stream iotexapi.APIService_StreamLogsServer) error {
```
Comment (CoderZhi) [inline api/coreservice.go:695]: same here

File: api/coreservice.go [hunk 16]
```diff
@@ -721,7 +714,7 @@ func (api *CoreService) StreamLogs(in *iotexapi.LogsFilter, stream iotexapi.APIS
 }
 
 // GetElectionBuckets returns the native election buckets.
-func (api *CoreService) GetElectionBuckets(epochNum uint64) ([]*iotextypes.ElectionBucket, error) {
+func (api *coreService) GetElectionBuckets(epochNum uint64) ([]*iotextypes.ElectionBucket, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:717]: ElectionBuckets

File: api/coreservice.go [hunk 17]
```diff
@@ -748,7 +741,7 @@ func (api *CoreService) GetElectionBuckets(epochNum uint64) ([]*iotextypes.Elect
 }
 
 // GetReceiptByActionHash returns receipt by action hash
-func (api *CoreService) GetReceiptByActionHash(h hash.Hash256) (*action.Receipt, error) {
+func (api *coreService) GetReceiptByActionHash(h hash.Hash256) (*action.Receipt, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:744]: ReceiptByActionHash

File: api/coreservice.go [hunk 18]
```diff
@@ -761,7 +754,7 @@ func (api *CoreService) GetReceiptByActionHash(h hash.Hash256) (*action.Receipt,
 }
 
 // GetActionByActionHash returns action by action hash
-func (api *CoreService) GetActionByActionHash(h hash.Hash256) (action.SealedEnvelope, error) {
+func (api *coreService) GetActionByActionHash(h hash.Hash256) (action.SealedEnvelope, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:757]: ActionByHash

File: api/coreservice.go [hunk 19]
```diff
@@ -771,7 +764,7 @@ func (api *CoreService) GetActionByActionHash(h hash.Hash256) (action.SealedEnve
 }
 
 // GetTransactionLogByActionHash returns transaction log by action hash
-func (api *CoreService) GetTransactionLogByActionHash(actHash string) (*iotextypes.TransactionLog, error) {
+func (api *coreService) GetTransactionLogByActionHash(actHash string) (*iotextypes.TransactionLog, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:767]: TransactionLogByActionHash

File: api/coreservice.go [hunk 20]
```diff
@@ -809,7 +802,7 @@ func (api *CoreService) GetTransactionLogByActionHash(actHash string) (*iotextyp
 }
 
 // GetTransactionLogByBlockHeight returns transaction log by block height
-func (api *CoreService) GetTransactionLogByBlockHeight(blockHeight uint64) (*iotextypes.BlockIdentifier, *iotextypes.TransactionLogs, error) {
+func (api *coreService) GetTransactionLogByBlockHeight(blockHeight uint64) (*iotextypes.BlockIdentifier, *iotextypes.TransactionLogs, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:805]: TransactionLogByBlockHeight or TransactionLogsByHeight

File: api/coreservice.go [hunk 21]
```diff
@@ -937,7 +930,7 @@ func (api *CoreService) getActionsFromIndex(totalActions, start, count uint64) (
 }
 
 // GetActions returns actions within the range
-func (api *CoreService) GetActions(start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
+func (api *coreService) GetActions(start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:933]: Actions
and we need to update return type as well in follow up PR

File: api/coreservice.go [hunk 22]
```diff
@@ -990,7 +983,7 @@ func (api *CoreService) GetActions(start uint64, count uint64) ([]*iotexapi.Acti
 }
 
 // getSingleAction returns action by action hash
-func (api *CoreService) GetSingleAction(actionHash string, checkPending bool) ([]*iotexapi.ActionInfo, error) {
+func (api *coreService) GetSingleAction(actionHash string, checkPending bool) ([]*iotexapi.ActionInfo, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:986]: Action

File: api/coreservice.go [hunk 23]
```diff
@@ -1003,7 +996,7 @@ func (api *CoreService) GetSingleAction(actionHash string, checkPending bool) ([
 }
 
 // getActionsByAddress returns all actions associated with an address
-func (api *CoreService) GetActionsByAddress(addrStr string, start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
+func (api *coreService) GetActionsByAddress(addrStr string, start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:999]: ActionsByAddress

File: api/coreservice.go [hunk 24]
```diff
@@ -1061,7 +1054,7 @@ func (api *CoreService) getActionByActionHash(h hash.Hash256) (action.SealedEnve
 }
 
 // getUnconfirmedActionsByAddress returns all unconfirmed actions in actpool associated with an address
-func (api *CoreService) GetUnconfirmedActionsByAddress(address string, start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
+func (api *coreService) GetUnconfirmedActionsByAddress(address string, start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:1057]: ditto

File: api/coreservice.go [hunk 25]
```diff
@@ -1087,7 +1080,7 @@ func (api *CoreService) GetUnconfirmedActionsByAddress(address string, start uin
 }
 
 // getActionsByBlock returns all actions in a block
-func (api *CoreService) GetActionsByBlock(blkHash string, start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
+func (api *coreService) GetActionsByBlock(blkHash string, start uint64, count uint64) ([]*iotexapi.ActionInfo, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:1083]: ditto

File: api/coreservice.go [hunk 26]
```diff
@@ -1111,7 +1104,7 @@ func (api *CoreService) GetActionsByBlock(blkHash string, start uint64, count ui
 }
 
 // getBlockMetas returns blockmetas response within the height range
-func (api *CoreService) GetBlockMetas(start uint64, count uint64) ([]*iotextypes.BlockMeta, error) {
+func (api *coreService) GetBlockMetas(start uint64, count uint64) ([]*iotextypes.BlockMeta, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:1107]: ditto

File: api/coreservice.go [hunk 27]
```diff
@@ -1136,7 +1129,7 @@ func (api *CoreService) GetBlockMetas(start uint64, count uint64) ([]*iotextypes
 }
 
 // getBlockMeta returns blockmetas response by block hash
-func (api *CoreService) GetBlockMetaByHash(blkHash string) ([]*iotextypes.BlockMeta, error) {
+func (api *coreService) GetBlockMetaByHash(blkHash string) ([]*iotextypes.BlockMeta, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:1132]: ditto

File: api/coreservice.go [hunk 28]
```diff
@@ -1582,7 +1573,7 @@ func (api *CoreService) getProtocolAccount(ctx context.Context, addr string) (*i
 }
 
 // GetActPoolActions returns the all Transaction Identifiers in the mempool
-func (api *CoreService) GetActPoolActions(actHashes []string) ([]*iotextypes.Action, error) {
+func (api *coreService) GetActPoolActions(actHashes []string) ([]*iotextypes.Action, error) {
```
Comment (CoderZhi) [inline api/coreservice.go:1576]: ditto

File: api/grpcserver.go [hunk 29]
```diff
@@ -87,14 +92,13 @@ func (svr *GrpcServer) SuggestGasPrice(ctx context.Context, in *iotexapi.Suggest
...
+	accountMeta, blockIdentifier, err := svr.coreService.Account(in.Address)
 	if err != nil {
 		return nil, err
 	}
 	return &iotexapi.GetAccountResponse{
-			AccountMeta:     accountMeta,
-			BlockIdentifier: blockIdentifier},
-		nil
+		AccountMeta:     accountMeta,
+		BlockIdentifier: blockIdentifier}, nil
```
Comment (CoderZhi) [inline api/grpcserver.go:101]:
```
    BlockIdentifier: blockIdentifier,
}, nil
```

File: api/coreservice.go [hunk 30]
```diff
@@ -0,0 +1,1600 @@
...
+// 	}
+// }
+
+// // WithNativeElection is the option to return native election data through API.
+// func WithNativeElection(committee committee.Committee) Option {
+// 	return func(cfg *Config) error {
+// 		cfg.electionCommittee = committee
+// 		return nil
+// 	}
+// }
```
Comment (dustinxie) [inline api/coreservice.go:92]: the above commented need to move to `serverV2.go`

Comment (dustinxie) [reply api/coreservice.go:92]: delete these
once we switch to v2, move from `api.go`

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**